cuqi/density/_density.py

```diff
@@ -0,0 +1,113 @@
+from __future__ import annotations
+from abc import ABC, abstractmethod
+from typing import Optional, Union
+
+import cuqi #  For type hints on conditioning
+class Density(ABC):
+    """ Abstract base class for densities.
+
+    Exposes a unified interface for evaluating the log density
+    depending on the parameters of the density. Common subclasses
+    are Distribution and Likelihood.
+
+    Parameters
+    ----------
+    name: str
+        Name of the random variable associated with the density.
+
+    Notes
+    -----
+    Subclasses must implement the following:
+
+    - _logd(self, *args, **kwargs). Returns the log density given a set of parameters.
+    - _condition(self, *args, **kwargs). This method conditions (fixes) the density on a given set of parameters.
+    - dim. The dimension of the density.
+    - get_parameter_names(self). Returns a list of the names of the parameters of the density.
+
+    """
+    def __init__(self, name: Optional[str] = None):
+        if not isinstance(name, str) and name is not None:
+            raise ValueError("Name must be a string or None")
+        self.name = name
+        self._constant = 0 # Precomputed constant to add to the log probability.
+
+    def logd(self, *args, **kwargs):
+        """ Evaluates the log density given a set of parameters.
+        
+        The log density can be evaluated with a set of arguments matching the parameter names
+        of the density. These can be accessed with the :meth:`get_parameter_names` method.
+        """
+
+        # Get parameter names possible to evaluate the logd
+        par_names = self.get_parameter_names()
+
+        # Check if kwargs are given. If so parse them according to the parameter names and add them to args.
+        if len(kwargs) > 0:
+
+            if len(args) > 0:
+                raise ValueError("Density.logd: Cannot specify both positional and keyword arguments.")
```

Comment:
Thanks. I added a way to get this name without using the inspect module. The inspect module is generally a bit slow and I am afraid if we do try excepts that this will slow down those parts of the code.